File: Trees/Binary Tree/Cousins in binary Tree/code.py
import queue


class Solution:
    def isCousins(self, root: TreeNode, x: int, y: int) -> bool:
        q = queue.Queue()
        q.put(root)
        while not q.empty():
            size = q.qsize()
            x_found = False
            y_found = False
            for i in range(size):
                curr = q.get()
                if curr.val == x:
                    x_found = True
                if curr.val == y:
                    y_found = True
                if curr.left != None and curr.right != None:
                    if (curr.left.val == x and curr.right.val == y) or (curr.left.val == y and curr.right.val == x):
                        return False
                if curr.left != None:
                    q.put(curr.left)
                if curr.right != None:
                    q.put(curr.right)

            if x_found and y_found:
                return True
            if x_found or y_found:
                return False

        return False

File: Trees/Binary Tree/Cousins in binary Tree/code.cpp
#include <bits/stdc++.h>
using namespace std;


class Solution {
public:
    bool isCousins(TreeNode* root, int X, int Y) {
        
        if(!root) return true;
        queue<TreeNode*>q;
        q.push(root);
        bool flag = false;
       
        while(!q.empty()){
            
        bool XFound = false;
        bool YFound = false;
            
            int levelSize = q.size();
            
            for( int i =0; i<levelSize; i++){
                
                TreeNode* currNode = q.front();
                q.pop();
                
                if(currNode->val == X) XFound = true;
                if(currNode->val == Y) YFound = true;

                
                if( currNode->left && currNode->right){
                    if((currNode->left->val == X && currNode->right->val == Y) || (currNode->left->val == Y && currNode->right->val == X)) return false;
                }
                
                if(currNode->left) q.push(currNode->left);
                if(currNode->right)q.push(currNode->right);   
            }
            
            if(XFound || YFound) flag = false;
            if( XFound && YFound) flag = true;
    
        } 
    return flag;
        
    }
}; 